R/write.R

```diff
@@ -5,139 +5,205 @@
+#' @importFrom dplyr select pull
+#' @importFrom rlang check_installed
 #'
 #' @return
-#' A workbook object
+#' An `openxlsx2` workbook object
 #'
 #' @examples
 #' \dontrun{
 #' redcap_uri <- Sys.getenv("REDCAP_URI")
 #' token <- Sys.getenv("REDCAP_TOKEN")
 #'
 #' supertbl <- read_redcap(redcap_uri, token)
-#' write_supertibble_xlsx(supertbl, file = "supertibble.xlsx")
+#' write_redcap_xlsx(supertbl, file = "supertibble.xlsx")
 #'
-#' write_supertibble_xlsx(supertbl, labelled = TRUE, file = "supertibble.xlsx")
+#' write_redcap_xlsx(supertbl, labelled = TRUE, file = "supertibble.xlsx")
 #' }
 #'
 #' @export
 
-write_supertibble_xlsx <- function(supertbl,
-                                   labelled = FALSE,
-                                   labelled_sheets = FALSE,
-                                   file,
-                                   incl_supertbl = TRUE,
-                                   incl_meta = TRUE,
-                                   tableStyle = "TableStyleLight10" #nolint: object_name_linter
+write_redcap_xlsx <- function(supertbl,
+                              file,
+                              labelled = NULL,
+                              use_labels_for_sheet_names = TRUE,
+                              include_toc_from_supertbl = TRUE,
+                              include_metadata = TRUE,
+                              table_style = "tableStyleLight8",
+                              set_col_widths = "auto",
+                              recode_yn = TRUE,
+                              na_strings = openxlsx2::na_strings()
 ) {
+  # Enforce checks ----
+  check_arg_is_supertbl(supertbl)
+  check_arg_is_character(table_style, any.missing = FALSE)
+  check_arg_is_character(file, any.missing = FALSE)
+  check_arg_is_valid_extension(file, valid_extensions = c("xlsx"))
+  check_arg_is_character(set_col_widths, any.missing = FALSE)
+  check_arg_is_logical(labelled, null.ok = TRUE)
+  check_arg_is_logical(use_labels_for_sheet_names, any.missing = FALSE)
+  check_arg_is_logical(include_toc_from_supertbl, any.missing = FALSE)
+  check_arg_is_logical(include_metadata, any.missing = FALSE)
+  check_arg_is_logical(recode_yn, any.missing = FALSE)
+
+  labelled <- check_labelled(supertbl, labelled)
+
   # Initialize Workbook object ----
-  wb <- openxlsx::createWorkbook()
+  check_installed("openxlsx2", reason = "to use `write_redcap_xl()`")
+  wb <- openxlsx2::wb_workbook()
 
   # Create Sheet Names ----
   # Assign sheet values based on use of labels
   # Enforce max length of 31 per Excel restrictions
-  sheet_vals <- if (labelled_sheets) {
-    supertbl$redcap_form_label
+  sheet_vals <- if (use_labels_for_sheet_names) {
+    # Remove special characters from labelled sheet names that cause
+    # openxlsx2 worksheet failures
+    supertbl$redcap_form_label %>%
+      str_replace_all("[[:punct:]]", "") %>%
+      str_squish()
   } else {
     supertbl$redcap_form_name
```

Comment:
Do REDCap form names never have any special characters that can cause `openxlsx2` worksheet failures?

Reply:
Unsure, this was because Excel sheets themselves can't contain special characters. You will make a file, but you'll be unable to open it/it will open with repairs. Same with the character limit of 31. There are some [additional rules](https://www.keynotesupport.com/excel-basics/worksheet-names-characters-allowed-prohibited.shtml) we may want to consider:

- Sheets can't be named "history"/"History"
- Punctuation is currently removed, but technically it can be used if it's wrapped in single quotes (personally I think it's better to just remove)
- Worksheets can't be blank